extensions/wikia/UserLogin/UserLoginSpecialController.class.php

```diff
@@ -257,72 +257,87 @@ public function getCloseAccountRedirectUrl() {
 	}
 
 	/**
-	 * @brief renders html version that will be inserted into ajax based login interaction
-	 * @details
-	 *   on GET, template partial for an ajax element will render
+	 * Renders html version that will be inserted into ajax based login interaction.
+	 * On GET, template partial for an ajax element will render
 	 */
 	public function dropdown() {
-		$query = $this->app->wg->Request->getValues();
-
-		if ( isset( $query['password'] ) ) {
-			// remove the password from the params to prevent exposiong in into the URL
-			unset($query['password']);
-		}
+		$query = $this->getSanitizedQueryString();
 
-		$this->returnto = $this->getReturnToFromQuery( $query );
-		$this->returntoquery = $this->getReturnToQueryFromQuery( $query );
+		$this->response->setData( [
+			'returnto' => $this->getReturnToFromQuery( $query ),
+			'returntoquery' => $this->getReturnToQueryFromQuery( $query ),
+		] );
 
 		$requestParams = $this->getRequest()->getParams();
 		if ( !empty( $requestParams[ 'registerLink' ] ) ) {
-			$this->registerLink = $requestParams[ 'registerLink' ];
+			$this->response->setVal( 'registerLink',  $requestParams[ 'registerLink' ] );
 		}
+
 		if ( !empty( $requestParams[ 'template' ] ) ) {
 			$this->overrideTemplate( $requestParams[ 'template' ] );
 		}
 	}
 
-	public function getMainPagePartialUrl() {
-		return Title::newMainPage()->getPartialURL();
-	}
-
 	/**
-	 * @param String $title
-	 *
-	 * @return Boolean
+	 * This method returns an array of query parameters that do not contain sensitive information
+	 * This can be used to construct returnto and returntoquery values that are not compromising.
 	 */
-	public function isTitleBlacklisted( $title ) {
-		return AccountNavigationController::isBlacklisted( $title );
+	private function getSanitizedQueryString() {
+		$query = $this->app->wg->Request->getValues();
+		$filterParams = [ 'password', 'username', 'loginToken' ];
+
+		foreach ( $filterParams as $paramName ) {
+			unset( $query[$paramName] );
+		}
+
+		return $query;
 	}
 
-	private function getReturnToFromQuery( $query ) {
-		if( !is_array( $query ) ) {
+	private function getReturnToFromQuery( &$query ) {
+		if ( !is_array( $query ) ) {
 			return '';
 		}
 
-		$returnto = $this->getMainPagePartialUrl();
-		if( isset( $query['title'] ) && !$this->isTitleBlacklisted( $query['title'] ) ) {
-			$returnto = $query['title'] ;
+		// If there's already a returnto here, use it.
+		if ( !empty( $query['returnto'] ) ) {
+			if ( !array_key_exists( 'returntoquery', $query ) ) {
```

Comment:
I feel like checking the value of `returnto` inside of `getReturnToQueryFromQuery` would be an alright approach. It seems better than setting a value inside of here, especially since that then couples these 2 functions so that `getReturnToFromQuery` will always have to be called before `getReturnToQueryFromQuery` to work properly.